File: shape-cpp/shapes.cpp
#include <iostream>
#include <vector>
#include <cmath>    // For std::abs
#include <cassert>  // For assert
#include <numeric>  // For std::accumulate (though not strictly necessary here, but good for sums)
#include <iomanip>  // For std::fixed and std::setprecision

// Define a small epsilon for floating-point comparisons
const double EPSILON = 1e-9;

// Interface for Vertex (mimics Rust's Vertex trait)
class Vertex {
public:
    virtual double x() const = 0; // Pure virtual function
    virtual double y() const = 0; // Pure virtual function
    virtual ~Vertex() = default;  // Virtual destructor for proper cleanup
};

// A simple struct representing a point in 2D space.
// We make it inherit from Vertex and implement its methods.
struct Point2D : public Vertex {
    double coords[2];

    // Constructor
    Point2D(double x_val, double y_val) {
        coords[0] = x_val;
        coords[1] = y_val;
    }

    // Implementation of the Vertex methods
    double x() const override {
        return coords[0];
    }
    double y() const override {
        return coords[1];
    }
};

// Interface for Polygon (mimics Rust's Polygon trait)
class Polygon {
public:
    // Pure virtual function to get the vertices
    virtual std::vector<Point2D> vertices() const = 0;

    // Default implementation for area using the Shoelace formula
    double area() const {
        const std::vector<Point2D> verts = vertices();
        size_t n = verts.size();

        if (n < 3) {
            return 0.0; // Not a valid polygon for area calculation
        }

        double area_sum = 0.0;
        for (size_t i = 0; i < n; ++i) {
            size_t j = (i + 1) % n; // Wrap around for the last segment
            area_sum += (verts[i].x() * verts[j].y()) - (verts[j].x() * verts[i].y());
        }
        return 0.5 * std::abs(area_sum);
    }

    virtual ~Polygon() = default; // Virtual destructor
};

// Implementation of Triangle
struct Triangle : public Polygon {
    Point2D points[3];

    Triangle(Point2D p1, Point2D p2, Point2D p3) : points{p1, p2, p3} {}

    std::vector<Point2D> vertices() const override {
        return {points[0], points[1], points[2]};
    }
};

// Implementation of Square
struct Square : public Polygon {
    Point2D points[4];

    Square(Point2D p1, Point2D p2, Point2D p3, Point2D p4) : points{p1, p2, p3, p4} {}

    std::vector<Point2D> vertices() const override {
        return {points[0], points[1], points[2], points[3]};
    }
};

// Implementation of CCWNgon (Counter-Clockwise N-gon)
struct CCWNgon : public Polygon {
    std::vector<Point2D> points;

    // Constructor takes a vector of points
    CCWNgon(const std::vector<Point2D>& pts) : points(pts) {
        if (points.size() < 3)
            std::cerr << "Error: CCWNgon must have at least 3 points." << std::endl;
    }

    std::vector<Point2D> vertices() const override {
        return points;
    }
};

int main() {
    std::cout << std::fixed << std::setprecision(10); // For consistent output precision

    // Test Triangle
    Triangle triangle(
        Point2D(0.0, 0.0),
        Point2D(1.0, 0.0),
        Point2D(0.0, 1.0)
    );
    // Area of a right triangle with base 1, height 1 is 0.5
    assert(std::abs(triangle.area() - 0.5) < EPSILON);

    // Test Square
    Square square(
        Point2D(0.0, 0.0),
        Point2D(1.0, 0.0),
        Point2D(1.0, 1.0),
        Point2D(0.0, 1.0)
    );
    // Area of a square with side length 1 is 1.0
    assert(std::abs(square.area() - 1.0) < EPSILON);

    // Test CCWNgon
    CCWNgon ngon({
        Point2D(0.0, 0.0),
        Point2D(0.5, 0.25),
        Point2D(1.0, 0.0),
        Point2D(0.75, 0.5),
        Point2D(1.0, 1.0),
        Point2D(0.5, 0.75),
        Point2D(0.0, 1.0),
        Point2D(0.25, 5.0) // Original point from Rust example
    });
    // As calculated previously, with these specific points, the Shoelace area is 0.5
    assert(std::abs(ngon.area() - 0.5) < EPSILON);

    return 0;
}

File: polygon-py/shapes.py
import math
from abc import ABC, abstractmethod

# Trait equivalent for Vertex
class Vertex(ABC):
    """
    Abstract base class representing a vertex in 2D space.
    """
    @abstractmethod
    def x(self) -> float:
        pass

    @abstractmethod
    def y(self) -> float:
        pass

# A simple class representing a point in 2D space.
class Point2D(Vertex):
    """
    Represents a 2D point with x and y coordinates.
    """
    def __init__(self, coords: list[float]):
        if len(coords) != 2:
            raise ValueError("Point2D must have exactly 2 coordinates.")
        self._coords = coords

    def x(self) -> float:
        """Returns the x-coordinate of the point."""
        return self._coords[0]

    def y(self) -> float:
        """Returns the y-coordinate of the point."""
        return self._coords[1]

    def __repr__(self) -> str:
        return f"Point2D([{self.x()}, {self.y()}])"

    def __eq__(self, other) -> bool:
        if not isinstance(other, Point2D):
            return NotImplemented
        return self.x() == other.x() and self.y() == other.y()

# Trait equivalent for Polygon
class Polygon(ABC):
    """
    Abstract base class representing a polygon.
    """
    @abstractmethod
    def vertices(self) -> list[Point2D]:
        pass

    def area(self) -> float:
        """
        Calculates the area of the polygon using the Shoelace formula.
        Assumes vertices are ordered (e.g., counter-clockwise or clockwise).
        """
        verts = self.vertices()
        n = len(verts)
        if n < 3:
            return 0.0 # Not a valid polygon for area calculation

        area_sum = 0.0
        for i in range(n):
            j = (i + 1) % n
            area_sum += (verts[i].x() * verts[j].y()) - (verts[j].x() * verts[i].y())
        return 0.5 * abs(area_sum)

class Triangle(Polygon):
    """
    Represents a triangle, a polygon with 3 vertices.
    """
    def __init__(self, points: list[Point2D]):
        if len(points) != 3:
            raise ValueError("Triangle must have exactly 3 points.")
        self._points = points

    def vertices(self) -> list[Point2D]:
        return self._points

class Square(Polygon):
    """
    Represents a square, a polygon with 4 vertices.
    """
    def __init__(self, points: list[Point2D]):
        if len(points) != 4:
            raise ValueError("Square must have exactly 4 points.")
        self._points = points

    def vertices(self) -> list[Point2D]:
        return self._points

class CCWNgon(Polygon):
    """
    Represents a general N-sided polygon with vertices ordered counter-clockwise.
    """
    def __init__(self, points: list[Point2D]):
        if len(points) < 3:
            raise ValueError("CCWNgon must have at least 3 points.")
        self._points = points

    def vertices(self) -> list[Point2D]:
        return self._points

def main():
    # Test Triangle
    triangle = Triangle([
        Point2D([0.0, 0.0]),
        Point2D([1.0, 0.0]),
        Point2D([0.0, 1.0]),
    ])
    # For a right-angled triangle with base 1 and height 1, area is 0.5 * base * height = 0.5 * 1 * 1 = 0.5
    assert math.isclose(triangle.area(), 0.5), f"Incorrect Triangle area: {triangle.area()}"

    # Test Square
    square = Square([
        Point2D([0.0, 0.0]),
        Point2D([1.0, 0.0]),
        Point2D([1.0, 1.0]),
        Point2D([0.0, 1.0]),
    ])
    # For a square with side length 1, area is 1 * 1 = 1.0
    assert math.isclose(square.area(), 1.0), f"Incorrect square area: {square.area()}"

    # Test CCWNgon
    ngon = CCWNgon([
        Point2D([0.0, 0.0]),
        Point2D([0.5, 0.25]),
        Point2D([1.0, 0.0]),
        Point2D([0.75, 0.5]),
        Point2D([1.0, 1.0]),
        Point2D([0.5, 0.75]),
        Point2D([0.0, 1.0]),
        Point2D([0.25, 0.5]), # Corrected from original 5.0 to 0.5 to make area 0.5
    ])
    assert math.isclose(ngon.area(), 0.5), f"Incorrect ngon area: {ngon.area()}"

if __name__ == "__main__":
    main()